src/Veldrid/StagingMemoryPool.cs

```diff
@@ -1,114 +1,133 @@
+            _storage = new List<StagingBlock>();
+            _availableBlocks = new SortedList<uint, uint>(new CapacityComparer());
+        }
 
-            return new StagingBlock(array, sizeInBytes, this);
+        public StagingBlock Stage(IntPtr source, uint sizeInBytes)
+        {
+            Rent(sizeInBytes, out StagingBlock block);
+            Unsafe.CopyBlock(block.Data, source.ToPointer(), sizeInBytes);
+            return block;
         }
 
         public StagingBlock Stage(byte[] bytes)
         {
-            byte[] array = _arrayPool.Rent(bytes.Length);
-            Array.Copy(bytes, array, bytes.Length);
-            return new StagingBlock(array, (uint)bytes.Length, this);
+            Rent((uint)bytes.Length, out StagingBlock block);
+            Marshal.Copy(bytes, 0, (IntPtr)block.Data, bytes.Length);
+            return block;
         }
 
-        public FixedStagingBlock GetFixedStagingBlock(uint sizeInBytes)
+        public StagingBlock GetStagingBlock(uint sizeInBytes)
         {
-            byte[] array = _arrayPool.Rent((int)sizeInBytes);
-            return new FixedStagingBlock(array, sizeInBytes, this);
+            Rent(sizeInBytes, out StagingBlock block);
+            Unsafe.InitBlock(block.Data, 0, sizeInBytes);
+            return block;
         }
 
-        public void Free(StagingBlock block)
+        public StagingBlock RetrieveById(uint id)
         {
-            bool clearArray = false;
-#if DEBUG
-            clearArray = true;
-#endif
-            _arrayPool.Return(block.Array, clearArray);
+            return _storage[(int)id];
         }
 
-        public void Free(FixedStagingBlock block)
+        private void Rent(uint size, out StagingBlock block)
         {
-            bool clearArray = false;
-#if DEBUG
-            clearArray = true;
-#endif
-            _arrayPool.Return(block.Array, clearArray);
+            lock (_lock)
+            {
+                SortedList<uint, uint> available = _availableBlocks;
+                IList<uint> indices = available.Values;
+                for (int i = 0; i < available.Count; i++)
+                {
+                    int index = (int)indices[i];
+                    StagingBlock current = _storage[index];
+                    if (current.Capacity >= size)
+                    {
+                        available.RemoveAt(i);
+                        current.SizeInBytes = size;
+                        block = current;
+                        _storage[index] = current;
+                        return;
+                    }
+                }
+
+                Allocate(size, out block);
+            }
         }
 
-        public void Free(byte[] array)
+        private void Allocate(uint sizeInBytes, out StagingBlock stagingBlock)
         {
-            bool clearArray = false;
-#if DEBUG
-            clearArray = true;
-#endif
-            _arrayPool.Return(array, clearArray);
+            IntPtr ptr = Marshal.AllocHGlobal((int)sizeInBytes);
```

Comment:
I think we should have a minimum allocation size. Perhaps 64 or 128 bytes.

Reply:
Seems reasonable.